---
title: "05_pyclone_annovar_cancer_gene_anno"
author: "Ming Tang"
date: "December 13, 2017"
output: html_document
---

```{bash}
cd results
mkdir pyclone_annotate_cancer_genes
```
### cancer gene list

```{r}
# cancer_genes<- read_tsv(here("data/cancer_genes/cancer_gene_kel.txt"), col_names = T)

# use cancer gene census from COSMIC!
cancer_genes<- read_tsv(here("data/COSMIC_Cancer_gene_Census/CGC_cancer_genes.tsv"), col_names = T)
cancer_genes<- dplyr::rename(cancer_genes, symbol = gene_symbol)

cancer_gene_mutations<- read_tsv(here("data/cancer_genes/wustl_curated_variants.tsv"), col_names = T,
                                 col_types = cols(.default = col_character()))

annovar_pyclone_files<- list.files("results/pyclone_annotate/", pattern = "_annovar_pyclone.tsv", full.names = T)

annovar_cancer_gene_anno<- function(x){
        oprefix<- gsub("_annovar_pyclone.tsv", "", basename(x))
        annovar_pyclone<- read_tsv(x, col_names = T, col_types = cols(.default = col_character()))

        annovar_pyclone_cancer_genes<- left_join(annovar_pyclone, cancer_gene_mutations, by = c("Chr" = "chromosome", "Start" = "start", "Ref" = "read", "Alt" = "variant")) %>%
                mutate(cancer_related_genes = ifelse(Gene.refGene %in% cancer_genes$symbol, 1, 0)) %>%
                dplyr::select(-starts_with("ExAC"), -starts_with("gnomAD"), -starts_with("CL", ignore.case = FALSE)) %>%
                mutate(cancer_gene_mutation = ifelse(is.na(gene), 0, 1 ))
        write_tsv(annovar_pyclone_cancer_genes, paste0("results/pyclone_annotate_cancer_genes/", oprefix, "_pyclone_cancer_genes.tsv"))
}

walk(annovar_pyclone_files, annovar_cancer_gene_anno )


```


```{r}
cadd_files<- list.files("results/pyclone_annotate_cancer_genes", pattern = "*tsv", full.names = T) 

cadd_dat_list <- lapply(cadd_files, function(f) {
        dat = read_tsv(f, col_names = T, col_types = cols(.default = col_character()))
        return(dat)
})

cadd_dat <- do.call(rbind, cadd_dat_list)

library(stringr)

## include indels, splicing, exon19 EGFR deletion. CADD > 20, cancer genes only, stop gain tumor suppressors.
## copy number alteration

cadd<- cadd_dat %>% dplyr::select(Chr, Start, End, Ref, Alt, 
                                  cosmic70, Gene.refGene, Func.refGene,ExonicFunc.refGene, CADD_phred, 
                                  AAChange.refGene, sample, cluster_id, cancer_related_genes,
                                  cancer_gene_mutation, variant_allele_frequency, cellular_prevalence) %>% 
        mutate(CADD_phred = as.numeric(CADD_phred)) %>%
        filter(Ref == "-" | Alt == "-"  | CADD_phred >= 20 | !is.na(cosmic70)) %>% 
        mutate(patient = str_replace(sample, "_T[1234]", "")) %>%
        arrange(desc(cancer_gene_mutation), cancer_related_genes, Gene.refGene) %>%
        filter(cancer_gene_mutation == "1" | cancer_related_genes == "1") %>%
        left_join(cancer_genes, by = c("Gene.refGene" = "symbol")) %>%
        group_by(patient) %>% nest() %>% arrange(patient)

walk(cadd$data, here("results", paste0(cadd$patient, "_mutations_CGC_filtered.txt")), write_tsv)

oncoprint_input<- cadd %>% unnest() %>% mutate(ExonicFunc.refGene = ifelse(is.na(ExonicFunc.refGene), Func.refGene, ExonicFunc.refGene )) %>% filter( ! ExonicFunc.refGene %in% c("synonymous SNV", "UTR3", "UTR5", "intronic"))  %>% filter( as.numeric(cellular_prevalence) >= 0.05)

## x is dataframe in long format.
turn_df_to_matrix<-function(x) {
        x<- x %>% spread(sample, value, fill = 0)
        m<-as.matrix(x[,-1])
        # x is a tibble, will not drop to vector
        rownames(m)<-x %>% pull(1)
        m
}

x<- oncoprint_list_dat$data[[8]] %>% spread(sample, value, fill = 0)

oncoprint_list_dat<- oncoprint_input %>% dplyr::select(Gene.refGene, ExonicFunc.refGene, sample) %>% 
        mutate(value =1) %>% 
        distinct() %>% 
        group_by(ExonicFunc.refGene) %>% 
        nest() 

 
oncoprint_mats<- oncoprint_list_dat %>% mutate(data_wd = map(.$data, turn_df_to_matrix))   

oncoprint_mats_list<- setNames(oncoprint_mats$data_wd, oncoprint_mats$ExonicFunc.refGene)

library(ComplexHeatmap)

mat_list<-  unify_mat_list(oncoprint_mats_list)

alter_fun = list(
    background = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "#CCCCCC", col = NA))
    },
    `nonsynonymous SNV` = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "blue", col = NA))
    },
    `nonframeshift deletion` = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "red", col = NA))
    },
    `frameshift deletion` = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "red3", col = NA))
    },
    `nonframeshift insertion` = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "green", col = NA))
    },
    `frameshift insertion` = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "green3", col = NA))
    },
    splicing = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.33, gp = gpar(fill = "#008000", col = NA))
    },
    stopgain = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.33, gp = gpar(fill = "yellow", col = NA))
    }
)

col<- c("nonsynonymous SNV" = "blue", "nonframeshift deletion"= "red", "frameshift deletion" = "red3",
        "nonframeshift insertion" = "green", "frameshift insertion" = "green3", 
        "splicing" = "#008000",  "stopgain" = "yellow")

oncoPrint(mat_list,
    alter_fun = alter_fun, col = col, 
    column_title = "mixing histology",
    column_order = NULL,
    show_column_names =T
    )

altered_nums<- Reduce( "+" , lapply(mat_list, rowSums))

NUMBER_GENES<- 15

slice_indx<- order(altered_nums, decreasing = TRUE)[1:NUMBER_GENES]

mat_list_sub<-  lapply(mat_list, function(x) x[slice_indx,])

oncoPrint(mat_list_sub,
    alter_fun = alter_fun, col = col, 
    column_title = "mixing histology",
    column_order = NULL,
    show_column_names =T
    )

heatmap_legend_param = list(title = "Alternations", at = c("AMP", "HOMDEL", "MUT"), 
        labels = c("Amplification", "Deep deletion", "Mutation"))


oncoPrint(mat_list_sub,
    alter_fun = alter_fun, col = col, 
    column_title = "mixing histology",
    row_order = NULL,
    show_column_names =T
    )
```


### only plot a list of genes given

```{r}
my_gene_list<- c("TP53", "EGFR", "RB1", "ATM", "PIK3CA", "DNMT3A", "NF1")

mat_list_sub2<-  lapply(mat_list, function(x) x[my_gene_list,])

oncoPrint(mat_list_sub2,
    alter_fun = alter_fun, col = col, 
    column_title = "mixing histology",
    # row_order = NULL,
    show_column_names =T
    )

```




